component/loki/relabel/relabel.go

```diff
@@ -0,0 +1,214 @@
+	cacheSize        prometheus_client.Gauge
+
+	cache *lru.Cache
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+
+	maxCacheSize = 10_000 // TODO(@tpaschalis) Do we make this configurable?
+)
+
+// New creates a new loki.relabel component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	cache, err := lru.New(maxCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:  o,
+		cache: cache,
+	}
+	c.entriesProcessed = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_processed",
+		Help: "Total number of log entries processed",
+	})
+	c.entriesOutgoing = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_written",
+		Help: "Total number of log entries forwarded",
+	})
+	c.cacheMisses = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_misses",
+		Help: "Total number of cache misses",
+	})
+	c.cacheHits = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_hits",
+		Help: "Total number of cache hits",
+	})
+	c.cacheSize = prometheus_client.NewGauge(prometheus_client.GaugeOpts{
+		Name: "loki_relabel_cache_size",
+		Help: "Total size of relabel cache",
+	})
+
+	for _, metric := range []prometheus_client.Collector{c.entriesProcessed, c.entriesOutgoing, c.cacheMisses, c.cacheHits, c.cacheSize} {
+		err := o.Registerer.Register(metric)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create and immediately export the receiver which remains the same for
+	// the component's lifetime.
+	c.receiver = make(loki.LogsReceiver)
+	o.OnStateChange(Exports{Receiver: c.receiver})
+
+	// Call to Update() to set the relabelling rules once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.receiver:
+			lbls := c.relabel(entry)
+			if len(lbls) == 0 {
+				level.Debug(c.opts.Logger).Log("msg", "dropping entry after relabeling", "labels", entry.Labels.String())
+				continue
+			}
+
+			c.entriesOutgoing.Inc()
+			entry.Labels = lbls
+			for _, f := range c.fanout {
+				select {
+				case <-ctx.Done():
+					return nil
+				case f <- entry:
+				}
+			}
+		}
+	}
+}
+
+func (c *Component) relabel(e loki.Entry) model.LabelSet {
+	c.entriesProcessed.Inc()
+	hash := e.Labels.Fingerprint().String()
+	found, ok := c.cache.Get(hash)
+	if ok {
+		c.cacheHits.Inc()
+		return found.(model.LabelSet)
+	}
+
+	c.cacheMisses.Inc()
+
+	// TODO(@tpaschalis) It's unfortunate how we have to cast back and forth
+	// between model.LabelSet (map) and labels.Labels (slice). Promtail does
+	// not have this issue as relabel config rules are only applied to targets.
+	// Do we want to use labels.Labels in loki.Entry instead?
```

Comment:
IMO we should switch everything to use labels.Labels as soon as we can. We should probably find out if there's a reason Promtail uses model.LabelSet, though.

Reply:
My idea was to get an end-to-end logging pipeline ready and then comparing if there's any big difference between the models.LabelSet and labels.Labels performance-wise, but I wouldn't think so for the number of labels usually in-use.

I'll reach out and see if there was a reason for this choice.